motionplan/constraint.go

```diff
@@ -10,120 +10,226 @@ import (
 	spatial "go.viam.com/rdk/spatialmath"
 )
 
-// ConstraintInput contains all the information a constraint needs to determine validity for a movement.
+// SegmentInput contains all the information a constraint needs to determine validity for a movement.
 // It contains the starting inputs, the ending inputs, corresponding poses, and the frame it refers to.
 // Pose fields may be empty, and may be filled in by a constraint that needs them.
-type ConstraintInput struct {
-	StartPos   spatial.Pose
-	EndPos     spatial.Pose
-	StartInput []referenceframe.Input
-	EndInput   []referenceframe.Input
-	Frame      referenceframe.Frame
+type SegmentInput struct {
+	StartPosition      spatial.Pose
+	EndPosition        spatial.Pose
+	StartConfiguration []referenceframe.Input
+	EndConfiguration   []referenceframe.Input
+	Frame              referenceframe.Frame
 }
 
-// Constraint defines functions able to determine whether or not a given position is valid.
-// TODO (pl): Determine how Gradient should fit into this
-// A bool returning whether the given input is known to be good, and a float representing how far the input is
-// from "ideal".
-type Constraint func(*ConstraintInput) (bool, float64)
+// Given a constraint input with only frames and input positions, calculates the corresponding poses as needed.
+func (ci *SegmentInput) resolveInputsToPositions() error {
```

Comment:
The current implementation is a *significant* efficiency speedup over what you're suggesting, as it allows cacheing of the resulting Position. Otherwise we're either 1) calculating the position needlessly, or 2) recalculating the same exact position many times over.

Reply:
I think I'm suggesting the same thing here except instead of a method on segment input it is a helper function that we use to set the segmentInput.  Why would this be slower?

Reply:
Oh I see; I think I misunderstood your first comment. Yes, that sounds fine.